webrender/src/renderer.rs

```diff
@@ -94,10 +94,70 @@ impl VertexDataTexture {
     }
 }
 
-#[derive(Copy, Clone)]
+const TRANSFORM_FEATURE: &'static [&'static str] = &["TRANSFORM"];
+
+enum ShaderKind {
+    Primitive,
+    Clear,
+}
+
+struct LazilyCompiledShader {
+    id: Option<ProgramId>,
+    name: &'static str,
+    kind: ShaderKind,
+    max_ubo_vectors: usize,
+    features: &'static [&'static str],
+}
+
+impl LazilyCompiledShader {
+    fn new(kind: ShaderKind,
+           name: &'static str,
+           max_ubo_vectors: usize,
+           features: &'static [&'static str],
+           device: &mut Device,
+           precache: bool) -> LazilyCompiledShader {
+        let mut shader = LazilyCompiledShader {
+            id: None,
+            name: name,
+            kind: kind,
+            max_ubo_vectors: max_ubo_vectors,
+            features: features,
+        };
+
+        if precache {
+            shader.get(device);
+        }
+
+        shader
+    }
+
+    fn get(&mut self, device: &mut Device) -> ProgramId {
+        match self.id {
+            Some(id) => id,
+            None => {
+                let id = match self.kind {
+                    ShaderKind::Clear => {
+                        create_clear_shader(self.name,
+                                            device,
+                                            self.max_ubo_vectors)
+                    }
+                    ShaderKind::Primitive => {
+                        create_prim_shader(self.name,
+                                           device,
+                                           self.max_ubo_vectors,
+                                           self.features)
+                    }
+                };
+                self.id = Some(id);
+                id
```

Comment:
How about putting this at the top of the function as:

    if self.id.is_none() {
        ...
    }

And then just return `self.id.unwrap()` out of this function.